src/rng.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * rng.cpp
+ *
+ *  Created on: 25 Mar 2020
+ *      Author: gc1610
+ */
+
+#include <Rcpp.h>
+
+float genexp(float av) {
+	return R::rexp(av);
+}
+
+float gennor(float av, float sd) {
+	return R::rnorm(av, sd);
+}
+
+float genunf(float low, float high) {
+	return R::runif(low, high);
+}
+
+
+void setgmn(float *meanv,float *covm,long p,float *parm)
+/*
+**********************************************************************
+     void setgmn(float *meanv,float *covm,long p,float *parm)
+            SET Generate Multivariate Normal random deviate
+                              Function
+      Places P, MEANV, and the Cholesky factoriztion of COVM
+      in GENMN.
+                              Arguments
+     meanv --> Mean vector of multivariate normal distribution.
+     covm   <--> (Input) Covariance   matrix    of  the  multivariate
+                 normal distribution
+                 (Output) Destroyed on output
+     p     --> Dimension of the normal, or length of MEANV.
+     parm <-- Array of parameters needed to generate multivariate norma
+                deviates (P, MEANV and Cholesky decomposition of
+                COVM).
+                1 : 1                - P
+                2 : P + 1            - MEANV
+                P+2 : P*(P+3)/2 + 1  - Cholesky decomposition of COVM
+               Needed dimension is (p*(p+3)/2 + 1)
+**********************************************************************
+*/
+{
+extern void spofa(float *a,long lda,long n,long *info);
+static long T1;
+static long i,icount,info,j,D2,D3,D4,D5;
+    T1 = p*(p+3)/2+1;
+/*
+     TEST THE INPUT
+*/
+    if(!(p <= 0)) goto S10;
+    Rcpp::Rcerr << "Value of P: " << p << std::endl;
+    Rcpp::stop("P nonpositive in SETGMN");
+S10:
+    *parm = p;
+/*
+     PUT P AND MEANV INTO PARM
+*/
+    for(i=2,D2=1,D3=(p+1-i+D2)/D2; D3>0; D3--,i+=D2) *(parm+i-1) = *(meanv+i-2);
+/*
+      Cholesky decomposition to find A s.t. trans(A)*(A) = COVM
```

Comment:
In the model, is it the case that A is constant across many runs? If so it will be potentially quite a bit faster (of lots of multivariate normal draws are done) to compute this only when needed - I've seen that before where multivariate normals were a big part of the simulation, though that was with an eigendecomposition of the VCV; that might be more expensive?